File: c++/leetcode/872. Leaf-Similar Trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
  void similar(TreeNode *, string &);
  bool leafSimilar(TreeNode *root1, TreeNode *root2)
  {

    string first = "";
    string second = "";

    similar(root1, first);
    similar(root2, second);

    return first == second;
  }
};

void Solution::similar(TreeNode *root, string &str)
{
  if (!root)
    return;

  if (!root->left && !root->right)
    str += root->val;

  similar(root->left, str);
  similar(root->right, str);
}

File: javascript/leetcode/872. Leaf-Similar Trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function (root1, root2) {
  let first = [];
  let second = [];

  traverse(root1, first);
  traverse(root2, second);

  console.log(first, "---", second);

  return (
    first.every((el, i) => el === second[i]) && first.length === second.length
  );
};

function traverse(root, arr) {
  if (!root) return;

  if (!root.left && !root.right) arr.push(root.val);

  traverse(root.left, arr);
  traverse(root.right, arr);
}
